Server/IO/GenericWriter.cs

```diff
@@ -0,0 +1,248 @@
+			{
+				Write(Serial.MinusOne);
+			}
+			else
+			{
+				Write(entity.Serial);
+			}
+		}
+
+		private void WriteEntity(ICustomsEntity entity)
+		{
+			if (entity == null || entity.Deleted)
+			{
+				Write(Serial.MinusOne);
+			}
+			else
+			{
+				Write(entity.Serial);
+			}
+		}
+
+		public virtual void Write(Point3D value)
+		{
+			Write(value.m_X);
+			Write(value.m_Y);
+			Write(value.m_Z);
+		}
+
+		public virtual void Write(Point2D value)
+		{
+			Write(value.m_X);
+			Write(value.m_Y);
+		}
+
+		public virtual void Write(Rectangle2D value)
+		{
+			Write(value.Start);
+			Write(value.End);
+		}
+
+		public virtual void Write(Rectangle3D value)
+		{
+			Write(value.Start);
+			Write(value.End);
+		}
+
+		public virtual void Write(Map value)
+		{
+			if (value != null)
+			{
+				Write((byte)value.MapIndex);
+			}
+			else
+			{
+				Write((byte)0xFF);
+			}
+		}
+
+		public virtual void Write(Item value) => WriteEntity(value);
+		public virtual void Write(Mobile value) => WriteEntity(value);
+		public virtual void Write(BaseGuild value) => Write(value?.Id ?? 0);
+		public virtual void Write(SaveData value) => WriteEntity(value);
+		public virtual void WriteItem<T>(T value) where T : Item => Write(value);
+		public virtual void WriteMobile<T>(T value) where T : Mobile => Write(value);
+		public virtual void WriteGuild<T>(T value) where T : BaseGuild => Write(value);
+		public virtual void WriteData<T>(T value) where T : SaveData => Write(value);
+
+		public virtual void Write(Race value)
+		{
+			if (value != null)
+			{
+				Write((byte)value.RaceIndex);
+			}
+			else
+			{
+				Write((byte)0xFF);
+			}
+		}
+
+        protected virtual void WriteListTidy(IList list, Action writerAction, bool tidy = false, Func<int, bool> tidyPredicate = null)
+		{
+			if (tidy && tidyPredicate == null)
+			{
+				throw new ArgumentNullException(nameof(tidyPredicate), "tidyPredicate must not be null if tidy is true.");
+			}
+
+			if (tidy)
+			{
+				for (var i = 0; i < list.Count;)
+				{
+					if (tidyPredicate(i))
+					{
+						list.RemoveAt(i);
+					}
+					else
+					{
+						++i;
+					}
+				}
+			}
+
+			Write(list.Count);
+			writerAction();
+		}
+
+		protected virtual void WriteSetTidy<T>(HashSet<T> set, Action writerAction, bool tidy = false, Predicate<T> tidyPredicate = null)
+		{
+			if (tidy && tidyPredicate == null)
+			{
+				throw new ArgumentNullException(nameof(tidyPredicate), "tidyPredicate must not be null if tidy is true.");
+			}
+
+			if (tidy)
+			{
+				set.RemoveWhere(tidyPredicate);
+			}
+
+			Write(set.Count);
+			writerAction();
+		}
+
+        protected virtual void WriteCol<T>(IEnumerable<T> list, Action<T> writer)
+        {
+               foreach(var o in list)
+                   writer(o);
+        }
+        
+		public virtual void WriteItemList(ArrayList list, bool tidy = false) => 
+			WriteListTidy(list, () => WriteCol(list.OfType<Item>(), Write), tidy, i => ((Item) list[i]).Deleted);

```

Comment:
There must be a better way to handle this, you're going to increase world save write time exponentially because each tidy check has to traverse the source list via [indexing].

The tidy predicate also does not check if the object is null and will crash.
A null object should be considered equivalent to a deleted object and thusly removed from the list.